cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/leveldb/LevelDBTableService.java

```diff
@@ -194,47 +209,39 @@ public void ensureTableExists(String tableName) throws IOException {
       synchronized (tables) {
         db = tables.get(tableName);
         if (db == null) {
-          createTable(tableName);
+          db = getDB(tableName, true);
+          tables.put(tableName, db);
         }
       }
     }
   }
 
-  private DB openTable(String tableName) throws IOException {
+  private DB getDB(String tableName, boolean createIfMissing) throws IOException {
     String dbPath = getDBPath(basePath, tableName);
 
     Options options = new Options();
-    options.createIfMissing(false);
+    options.createIfMissing(createIfMissing);
     options.errorIfExists(false);
     options.comparator(new KeyValueDBComparator());
     options.blockSize(blockSize);
     options.cacheSize(cacheSize);
-
-    // unfortunately, with the java version of leveldb, with createIfMissing set to false, factory.open will
-    // see that there is no table and throw an exception, but it wont clean up after itself and will leave a
-    // directory there with a lock.  So we want to avoid calling open if the path doesn't already exist and
-    // throw the exception ourselves.
-    File dbDir = new File(dbPath);
-    if (!dbDir.exists()) {
-      throw new IOException("Database " + dbPath + " does not exist and the create if missing option is disabled");
+    options.maxOpenFiles(isSystemTable(tableName) ? systemTablesMaxOpenFiles : userTablesMaxOpenFiles);
+
+    if (!createIfMissing) {
+      // unfortunately, with the java version of leveldb, with createIfMissing set to false, factory.open will
+      // see that there is no table and throw an exception, but it wont clean up after itself and will leave a
+      // directory there with a lock.  So we want to avoid calling open if the path doesn't already exist and
+      // throw the exception ourselves.
+      File dbDir = new File(dbPath);
+      if (!dbDir.exists()) {
+        throw new IOException("Database " + dbPath + " does not exist and the create if missing option is disabled");
+      }
     }
-    DB db = factory.open(dbDir, options);
-    tables.put(tableName, db);
-    return db;
+    return factory.open(new File(dbPath), options);
   }
 
-  private void createTable(String name) throws IOException {
-    String dbPath = getDBPath(basePath, name);
-
-    Options options = new Options();
-    options.createIfMissing(true);
-    options.errorIfExists(false);
-    options.comparator(new KeyValueDBComparator());
-    options.blockSize(blockSize);
-    options.cacheSize(cacheSize);
-
-    DB db = factory.open(new File(dbPath), options);
-    tables.put(name, db);
+  private boolean isSystemTable(String tableName) {
+    return tableName.startsWith(String.format("%s_%s", tablePrefix, NamespaceId.SYSTEM.getNamespace()));
```

Comment:
aren't there some system tables in the user namespace? Perhaps rely on this method: https://github.com/caskdata/cdap/blob/75f73509d2e7fefb1aa2f646651311ba00cdbe70/cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/DatasetsUtil.java#L163

Reply:
Pushed a commit to extract the DatasetId from the table name. I was then able to use this.